[TensorExpr] Properly handle all dtypes in evaluation of Intrinsics exprs.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42495 [TensorExpr] Properly handle all dtypes of the condition in evaluation of IfThenElse exprs.
* **#42494 [TensorExpr] Properly handle all dtypes in evaluation of Intrinsics exprs.**
* #42493 [TensorExpr] Properly handle all dtypes in evaluation of CompareSelect exprs.

Note that we're currently assuming that dtypes of all the arguments and
the return value is the same.

Differential Revision: [D22910755](https://our.internmc.facebook.com/intern/diff/D22910755)

Commit:
[TensorExpr] Properly handle all dtypes in evaluation of Intrinsics exprs.

Note that we're currently assuming that dtypes of all the arguments and
the return value is the same.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Properly handle all dtypes in evaluation of Intrinsics exprs."


Note that we're currently assuming that dtypes of all the arguments and
the return value is the same.

Differential Revision: [D22910755](https://our.internmc.facebook.com/intern/diff/D22910755)

[ghstack-poisoned]

nickgg reviewed (commented):
This one's a bit weird, are all Intrinsics supported for all types? sin(bool) seems pretty odd for instance.

Commit:
Update on "[TensorExpr] Properly handle all dtypes in evaluation of Intrinsics exprs."


Note that we're currently assuming that dtypes of all the arguments and
the return value is the same.

Differential Revision: [D22910755](https://our.internmc.facebook.com/intern/diff/D22910755)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Properly handle all dtypes in evaluation of Intrinsics exprs."


Note that we're currently assuming that dtypes of all the arguments and
the return value is the same.

Differential Revision: [D22910755](https://our.internmc.facebook.com/intern/diff/D22910755)

[ghstack-poisoned]

ZolotukhinM commented:
> This one's a bit weird, are all Intrinsics supported for all types? sin(bool) seems pretty odd for instance.

Good point. I originally thought that we have intrinsics working with all the types but it looks like in reality we only have intrinsics working with `float` and `double`. I updated the PR to only allow these two dtypes.

nickgg reviewed (approved):
Works for me.
